### ISSUE TITLE
an 'or' step with two 'has' steps that test different property keys followed by an 'as' step results in nothing

### ISSUE THREAD

**Issue by caamartin35** [JanusGraph/janusgraph#1938]
I have been using JanusGraph for some time now and have come across something I believe to be an issue.

I would like someone to confirm this and let me know if it is an already known issue.

For all of the below queries I am using the GraphOfTheGods Graph.

If we do a regular 'or' step, we get the correct results:
```
gremlin> g.V().or(has("name", "neptune"), has("name", "pluto"))
==>v[4104]
==>v[8280]
```

If we change one of the keys to be different than the other, it still works as expected:
```
gremlin> g.V().or(has("name", "neptune"), has("age", 4000))
12:34:16 WARN  org.janusgraph.graphdb.transaction.StandardJanusGraphTx  - Query requires iterating over all vertices [(age = 4000)]. For better performance, use indexes
==>v[4104]
==>v[8280]
```

If we add an 'as' and a 'select' step to the original query it works as expected:
```
gremlin> g.V().or(has("name", "neptune"), has("name", "pluto")).as("v").select("v")
==>v[4104]
==>v[8280]
```

However, if we add 'as' step and a 'select' step to an 'or' step with two 'has' steps that are testing against different property keys, it always returns no results:
```
gremlin> g.V().or(has("name", "neptune"), has("age", 4000)).as("v").select("v")
12:37:35 WARN  org.janusgraph.graphdb.transaction.StandardJanusGraphTx  - Query requires iterating over all vertices [(age = 4000)]. For better performance, use indexes
```

If we add a step in after the 'or' step then it works as expected:
```
gremlin> g.V().or(has("name", "neptune"), has("age", 4000)).valueMap().as("v").select("v")
12:38:11 WARN  org.janusgraph.graphdb.transaction.StandardJanusGraphTx  - Query requires iterating over all vertices [(age = 4000)]. For better performance, use indexes
==>[name:[neptune],age:[4500]]
==>[name:[pluto],age:[4000]]
```

If we change the 'or' step to an 'and' step it works as expected:
```
gremlin> g.V().and(has("name", "neptune"), has("age", 4500)).as("v").select("v")
==>v[4104]
```

So it seems that having an 'or' step with two 'has' steps that test different property keys followed by an 'as' step results in nothing if that 'as' is selected at some point.

You can also note the warnings that print out in these different cases. The 'use indexes' warning prints for the or() with different keys case but not the and() case or the or() with the same key case. That seems odd to me.

I have tested this on both JanusGraph and TinkerGraph. It seems to only be an issue on JanusGraph, so it doesn't seem to be a TP/Gremlin issue, but instead a JG issue.

**Comment by vtslab** [comment]
Duplicate bug report, see [https://github.com/JanusGraph/janusgraph/issues/1714](https://github.com/JanusGraph/janusgraph/issues/1714)

**Comment by mad** [comment]
It will be fixed here https://github.com/JanusGraph/janusgraph/pull/1943

**Comment by FlorianHockmann** [comment]
Closed as a duplicate of #1714. Thanks for noting this @vtslab!
